### ISSUE TITLE
Enhancements - vertical video alignment, cycle modes and YT button

### ISSUE THREAD

**Issue by markusd1984** [tamius-han/ultrawidify#145]
First of huge thanks and hats up @tamius-han and team for this amazing extension!
It certainly offers the most settings and features from what I saw, comparing it to UltraVideo extension with only limited keyboard shortcuts and Zoom to Fill which only offers the simple convenient YT button. 

Some feedback and suggestions I would like to share that I think could make it even better if we could implement them (i can open separate issues but just wanted to get feedback first).

**1. Video Alignment to the Top / Bottom**
so videos can be cropped to 21:9 either from the top or bottom rather then only zoomed in and loosing on both top and bottom (or stretching), which can be less optimal for head shots or information that are in different positions, i.e. too far on the top or bottom displayed. Even better switching throughout the video could offer the most flexibility and best Ultrawide 21:9 experience with a lesser compromise, which I think is desperately needed, if that's actually possible to implement?

**2.  Shortcuts to cycle Modes** for Cropping, Stretch and Video Alignment
because often it's easier to remember one shortcut then many and also it would make trying different modes much easier and quicker, has this been considered already?

**3. Nice to have - YT button**
I understand the [YT button has been removed](https://github.com/tamius-han/ultrawidify/issues/11#issuecomment-354650779) a few years ago due to complexity and maintenance  required, any possibility this may be considered for coming back?

It's quite a nice convenience feature, having used Zoom to Fill extension, as sometimes it's easier when being lazy to use shortcuts and also some people don't like shortcuts. This extension could offer the most flexibility for a YT button in terms of cycling through different modes, since the extensive modes and custom (experimental) settings are already built in.

**Comment by tamius-han** [comment]
> **1. Video Alignment to the Top / Bottom**

Perhaps/eventually. Low priority because shift+[move mouse over video] already offers _some_ degree of ... that. (With some limitations)

> Even better switching throughout the video could offer the most flexibility and best Ultrawide 21:9 experience with a lesser compromise, which I think is desperately needed,

If you mean aspect ratio detection, that's already a thing. 

There are some limitations due to performance reasons (namely, if subtitles appear and are contained entirely within the black bar, extension will miss them). This is a thing that is unlikely to be ever fixed, as this is a very messy edge case.

> **2. Shortcuts to cycle Modes for Cropping, Stretch and Video Alignment**

That was not yet considered. Maybe if I ever get to rewriting keyboard shortcuts, since this isn't very compatible with how keyboard shortcuts are handled now.

(Though it's possible that keyboard shortcut handling rewrite may happen sooner than expected).

> 3. Nice to have - YT button

There are plans to have in-player UI in the future. So far, the progress is rather slow, but it looks like it's going to work (since I have managed to inject Vue components into webpages).

There are a few bumps, though — namely, the 'button' needs to be something that works everywhere. Also, getting vue and the rest of the extension to communicate is a bit iffy.

**Comment by markusd1984** [comment]
oh sweet, shift + mouse is even more user friendly/easier than i expected. 😄 👍

What I meant re switching is to have it either on the very top or bottom the crop/alignment,
which might still be worth while adding shortcuts instead of using shift and going all the way to the top/bottom.

Re cycling modes, would this require a rewriting though, could these not be simply added as an additional function, that could be carried over if you'd ever re-write it?

Even just for the crop mode which I think is what I would expect is the most frequent used feature, which may be changed for different videos, e.h. pressing ALT+C changes from automatic, Fit width, Fit Height, 21:9. 18:9, 16:9, stretch (in that order i would expect to be most common preference on ultrawide monitors for the best viewing experience).

Re button, makes sense. Maybe you can have a logic only on YT to show the button to start with, so you just need to make it work for the most popular video platform, and later add more support to others e.g. vimeo, amazon, netflix etc. but at least have support for YT. That also means there would be no reason to use other extensions for this and thus Ultrawidify will get more usage.

**Comment by tamius-han** [comment]
> What I meant re switching is to have it either on the very top or bottom the crop/alignment,
which might still be worth while adding shortcuts instead of using shift and going all the way to the top/bottom.

As I said, I've been thinking about it, but it's low priority.

> Re cycling modes, would this require a rewriting though, could these not be simply added as an additional function

If you're willing to resort to ugly hacks, then yes. However, I am not sure this would go over well with the settings page, so I think I'll leave that be for a moment. There's a chance I'll have to rewrite a few things about keyboard shortcuts when I get to having an in-page GUI, so currently cycling is not a priority until I get to that part. 

> that could be carried over if you'd ever re-write it?

The rewrite that might happen might be a little more complex than that. It might require rewriting keyboard handling from scratch.

**Comment by markusd1984** [comment]
Thanks appreciate that.

Re ug
> > What I meant re switching is to have it either on the very top or bottom the crop/alignment,
> which might still be worth while adding shortcuts instead of using shift and going all the way to the top/bottom.
> 
> As I said, I've been thinking about it, but it's low priority.

Just wanted to clarify that I did not meant aspect ratio detection you referred to. =)


> > Re cycling modes, would this require a rewriting though, could these not be simply added as an additional function
> 
> If you're willing to resort to ugly hacks, then yes. However, I am not sure this would go over well with the settings page, so I think I'll leave that be for a moment. There's a chance I'll have to rewrite a few things about keyboard shortcuts when I get to having an in-page GUI, so currently cycling is not a priority until I get to that part. 

Ugly or interruptive under the main 'quick settings' menu perhaps but this could be easily added under advanced options settings that is already full of a long list of additional settings, why not simply add it there for now, surely one more won't make a difference?

**Comment by markusd1984** [comment]
I discovered that you can add additional actions for the commands and also create new actions on the options page, that is really amazing the flexibility you already built in! @tamius-han  

Perhaps you could just add support for the **Top/Bottom alignment** i.e. "Top" using arg value 3 & "Bottom" using arg value 4, so we can select it from the list and this way create out own actions?

![image](https://user-images.githubusercontent.com/24621214/114366112-09a0a780-9b73-11eb-85c6-19ed657786ab.png)

**Comment by markusd1984** [comment]
FYI - re cycling option, maybe it could be as easy and dirty to add four options under "Other Actions":

- Cycle crop actions
- Cycle stretch actions
- Cycle alignment actions
- Cycle zoom / panning actions

That way even when new actions are added to the groups they can be cycle through.
Disabled by default so one can enable only when desired and set a short cut each.
Incl. the check mark to show the popup in video settings (page), which makes it easier to check what short cut has been set.

Or perhaps better even have just one entry "Cycle through group of actions" in which one can enable the group of actions to use and set each shortcut. 😄 
It looks like you have most of the framework for this already in use so perhaps could be fairly easily replicated for this except for the logic to cycle through the actions group?

**Comment by markusd1984** [comment]
@tamius-han  any update or more considerations? 

Just switched to mac and when using the extension again experience the need to use 21:9 but not have an equal crop top/bottom effect by just zooming in the centre.

**Would it be possible to add support for "set video alignment" to provide action parameter for Top & Bottom please?**

That way we can use hotkey to use the zoom in when beneficial but switching between top vs centre vs bottom alignment that will yield best utilisation based on content and help with cutting off peoples heads or information.

**Comment by tamius-han** [comment]
This is the plan for when I go around to finishing the move to in-player UI.

I honestly hoped I'd manage to finish the UI re-design by now, but for some reason I've had a severe lack of free time this year. I initially planned to work on that back in May or June ... but at current pace I'm gonna be lucky if I manage to pull this off by the end of the year.

**Comment by markusd1984** [comment]
Thanks for update, I understand and it's amazing you're investing time into it and freely share the results.

I just wish it had this flexibility as the main reason for using the extension is to utilize the full widescreen but not being in control which part (top/bottom) gets cut off limits this experience quite a lot and as a result, often I switch back to normal playback and don't use the plugin. I hoped it would have been a minor modification to add this parameter, once there is an update please let me know and I'm happy to help test it.

**Comment by markusd1984** [comment]
> having an in-page GUI, so currently cycling is not a priority until I get to that part.

HNY! Just wondering if you had any more time working on the re-design or perhaps adding this feature mod? 

This I believe is a game changer setting, given the point of ultrawidify is to crop video to suit ultrawide monitors thus vertical alignment is important too for usability.

**Comment by tamius-han** [comment]
> if you had any more time working on the re-design

Not nearly as much as I would have liked. (Also, there's been quite some annoyances along the way)

That being said, the option is already included in re-design, however the re-design isn't quite as close to feature parity with the old thing as I would like + some new features are still too half-baked to release.

The interface for the alignment is not _very_ stellar, but I'm open for comments regarding design:

![image](https://user-images.githubusercontent.com/12505802/211074039-be0dab9a-9c9c-47c8-8d3e-337cb4e282be.png)

**Comment by tamius-han** [comment]
Here's the current build to play with.  

Things done so far: 
* in-player UI works on Youtube and twitch. 
* communication between in-player UI, popup, and the rest of the scripts
* cycle on aspect ratio

Things that need to be done before release:
* feature parity on sites with iframes (e.g. /r/videos on reddit). Needed because current popup has this feature.
* while extension popup will be removed in the future, i want to support both in-player UI as well as popup to alleviate complaining that inadvertantly comes with a big change like this
* options to manually select the player and video elements from the in-page UI. Current status of the feature: present but broken (for player). "Quick fixes" column will probably vanish on release, as I don't think I'll manage to implement it for reals by end of may.
* I'm also changing the default behaviour of the extension. Current default is to always run, on every site. New default: run only when in full screen. (Youtube default: run in full screen and theater mode, twitch: always run)

Things that will probably be left out of the _next_ release, but will have to be taken care of by end of this year:
 * manifest v3
 
 # Downloads

[Ultrawidify-5.99.0-4-ff.zip](https://github.com/tamius-han/ultrawidify/files/11106218/Ultrawidify-5.99.0-4-ff.zip)
[Ultrawidify-5.99.04-chrome.zip](https://github.com/tamius-han/ultrawidify/files/11106219/Ultrawidify-5.99.04-chrome.zip)
[Ultrawidify-5.99.04-edge.zip](https://github.com/tamius-han/ultrawidify/files/11106220/Ultrawidify-5.99.04-edge.zip)

Load as unpacked extension (`about:debugging` in firefox, edge and chrome have `chrome://extensions/` — don't forget to enable developer mode ). Firefox requires you to load unpacked extension every time you launch the browser, Chrome allows you to load unpacked extension and run it indefinitely, Edge allows you to load unpacked extension and run it indefinitely, but it will nag you to disable unpacked extensions every now and then.

## Known issues

* Unpacked extension isn't going to work until you manually set alignment when you launch it for the first time

**Comment by markusd1984** [comment]
Thanks and sorry for late reply, only got to try it out now in Chrome. 

Unfortunately I only see white, it looks like the UI is overlaying the video. Hover over "aspect ratio control" brings up the settings and disappears when navigating the mouse outside the settings window leaving a white blank screen. Tried re-enabling and original installed extension is disabled. 

Am I missing something?

**Comment by markusd1984** [comment]
Just wondering if you're still working on this / if there is a newer build available to test? cheers

**Comment by markusd1984** [comment]
Just to check in a year later in case you happen to make any more progress on this new version? Any new ones to try?

**Comment by tamius-han** [comment]
Sorry for long time no response.

New version is pending review on chrome web store. Hopefully out by the end of the week.
